llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp

```diff
@@ -3855,11 +3855,19 @@ bool RISCVDAGToDAGISel::performCombineVMergeAndVOps(SDNode *N) {
   // If we end up changing the VL or mask of True, then we need to make sure it
   // doesn't raise any observable fp exceptions, since changing the active
   // elements will affect how fflags is set.
-  if (TrueVL != VL || !IsMasked)
+  if (TrueVL != VL || !IsMasked) {
     if (mayRaiseFPException(True.getNode()) &&
         !True->getFlags().hasNoFPExcept())
       return false;
 
+    // If the EEW of True is different from vmerge's SEW, then we cannot change
+    // the VL or mask.
+    if (Log2_64(True.getScalarValueSizeInBits()) !=
+        N->getConstantOperandVal(
+            RISCVII::getSEWOpNum(TII->get(N->getMachineOpcode())) - 1))
```

Comment:
The opcode is known? So I think we don't need to get SEW operand index from TSFlag, we can hardcode it here.